text/0000-disjoint-groups/0000-disjoint-groups.md

```diff
@@ -0,0 +1,149 @@
+
+If the destination is a _group authority_ with address `d`:
+
+* If `t` extends `p`, handle the message and relay it to everyone else in `(p)`.
+* Otherwise relay the message to the `route`-th closest entry to `t` in our routing table.
+
+For any groups `A` and `B`, either everyone in `A` is closer to `d` than everyone in `B`, or vice versa. Therefore, the _group_ that we relay the message to doesn't depend on the route. Since everyone in our group knows everyone in the next hop's group, this means that in the next attempt `route + 1`, a _different_ member of `B` will receive the message. Unless there is churn in between the attempts, this guarantees that all routes are disjoint.
+
+As mentioned above, the current group message routing mechanism is not secure. To prevent intercepting group messages, we will instead relay them from group to group, verifying quorum in every step:
+
+Assume a node `n` in group `B` receives a group message signed by someone in group `A`:
+
+* If `A` is not connected to `B`, drop the message.
+* If this message has not yet been signed by a quorum of members of `A`, keep it in the cache.
+* If it has been signed by a quorum of members of `A`: If we are the recipient, handle it. If not, find the connected group `C` closest to the destination and send the message, signed by `n`, to _every member_ of `C`.
+
+This creates a number of direct messages in each hop that grows quadratically in the group size. To reduce network traffic, the `route`-th node of group `B` could accumulate `A`'s signatures instead, making the message number linear in the group size, at the cost of having effectively two network hops per group hop:
+
+* If `A` is not connected to `B`, drop the message.
+* If this message has not yet been signed by a quorum of members of `A`, keep it in the cache.
+* If it has been signed by a quorum of members of `A`: If we are the recipient, handle it. If not, find the connected group `C` closest to the destination and send the message, signed by `n`, to the `route`-th closest member of `C`.
+* If it has been signed by a quorum of members of `A` and we are the `route`-th closest member of `B`, send the message with `A`'s signatures to every member of `B`.
+
+The quorum cannot be a constant anymore, due to varying group sizes. It needs to be a percentage strictly greater than 50% instead, and in a group of size `n`, a number `x` of nodes will constitute a quorum if `x / n >= QUORUM`.
+
+### Joining nodes
+
+A new node needs to first connect to everyone in its target group. After that, it can drop its proxy node. The other group members can provide the node with their routing tables, which should all be identical. It can then make connections to these other groups, too.
+
+### Group split
+
+If a group `(p)` splits, all its nodes stay connected: The new groups are now each other's `i`-th bucket, where `i` is the number of bits in `p`.
```

Comment:
Does it also disconnect from the previous connected group that agreed on the current leading bits - 1 ?